### DISCUSSION TITLE
how to include external resources file starting from version 3

### DISCUSSION THREAD

**Discussion by blurstream** [serverless/serverless#10665] [Q&A]
hello, i've updated to the latest serverless version (currently 3.1.1) but it seems this version isn't supporting external file inclusion anymore.
i searched through open issues and documentation but i haven't found nothing useful.
what i'm doing is trying to include external files (using different stages) under `resources` section:
```
resources:
  - Resources:
       ...
  - ${file(./resources/${self:provider.stage}.yml)}
```
under the same directory of `serverless.yml` file i had created resources folder with all available stages.
`${self:provider.stage}` is defined this way:
```
provider:
  name: aws
  runtime: go1.x
  stage: ${opt:stage, 'dev'}
```
but when sls deploys this configuration it raises the following error: 
```
Error:
--
Cannot resolve serverless.yml: Variables resolution errored with:
  - Cannot resolve variable at "resources.1": Value not found at "file" source
```
i can't figure out what's happening. could you give me any suggestions? thank you!

**Comment by mnapoli** [comment]
Sorry to not directly answer your question (maybe there's a bug here?), but I just wanted to point out that you can use the `${sls:stage}` variable to simplify all of that.

No need for this line anymore:

```
  stage: ${opt:stage, 'dev'}
```

`${sls:stage}` resolves to the current stage, so it's a little simpler to use.

**Reply by dbersan** [reply]
how is one supposed to define the default stage, in case it is not specified? `stage: ${opt:stage, 'dev'}` defined the default clearly, `${sls:stage}` does not

**Comment by blurstream** [comment]
thank you @mnapoli for the simplification!
after some tests, i found the issue seems related to the file content. if this is empty then that error is triggered.
i guess it's due to a change to variables resolution system, but i can't find references to what changed both in documentation and in commits. ideas? thank you

**Reply by mnapoli** [reply]
This may be related to the switch to the new variable resolver by default in v3: https://www.serverless.com/framework/docs/guides/upgrading-v3#low-level-changes

The new resolver is more robust and surfaces errors more strictly, so that's what comes to mind.

**Comment by najibla** [comment]
We are having the same issue. None of the files is empty, yet the error.
